src/rov/rov_pca9685/src/PCA9685.cpp

```diff
@@ -0,0 +1,190 @@
+#include "pca9685/PCA9685.hpp"
+
+extern "C" {
+#include "linux/i2c-dev.h"
+#include "i2c/smbus.h"
+}
+
+#include <cstddef>
+#include <sys/ioctl.h>
+#include <cstdlib>
+#include <cstdio>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <cmath>
+
+#define US_PER_S 1e6
+
+PCA9685::PCA9685(uint8_t bus, uint8_t address) : i2c_address(address), i2c_bus(bus) {
+    clock_frequency = 25e6;
+    // 1e6 us per s / 25e6 hz = 0.04 us
+    clock_us = US_PER_S / 25e6;
+    output_frequency = 200;
+}
+
+PCA9685::~PCA9685() {
+    Close();
+}
+
+bool PCA9685::Open() {
+    char filename[32];
+    sprintf(filename,"/dev/i2c-%d", i2c_bus);
+    // printf("%s\n", filename);
+    i2c_fd = open(filename, O_RDWR);
+    if (i2c_fd < 0) {
+        // Could not open the bus
+    
+        return false;
+    }
+    if (ioctl(i2c_fd, I2C_SLAVE, i2c_address) < 0) {
+        // Could not open the device on the bus
+        return false;
+    }
+
+    // external clock?
+    // i2c_smbus_write_byte_data(i2c_fd, MODE1, MODE1_SLEEP);
+    // usleep(500);
+    // i2c_smbus_write_byte_data(ic2_fd, MODE1, MODE1_SLEEP | MODE1_EXTCLK);
+
+    usleep(500);
+    setFrequency(60);
+    
+    // startup pattern
+    // for (uint8_t idx = 0; idx < 16; idx++) {
+    //     setOn(idx);
+    //     usleep(5000); 
+    //     setAllOff();
+    // }
+    // for (uint8_t idx = 14; idx < 16; idx--) {
+    //     setOn(idx);
+    //     usleep(5000); 
+    //     setAllOff();
+    // }
+
+    // clear all registers
+    Reset();
+
+    return true;
+}
+
+void PCA9685::Close() {
+    setAllOff();
+    usleep(500);
+    if (i2c_fd > 0) {
+        close(i2c_fd);
+    }
+}
+
+void PCA9685::Reset() {
+    // stop all pwms
+    i2c_smbus_write_byte_data(i2c_fd, Registers::ALL_LED_OFF_H, BIT4);
+    usleep(500);
+    
+    uint8_t mode1_current = i2c_smbus_read_byte_data(i2c_fd, Registers::MODE1);
+    if (mode1_current & MODE1_RESTART) {
+        i2c_smbus_write_byte_data(i2c_fd, Registers::MODE1, mode1_current & ~MODE1_SLEEP);
+        usleep(500);
+    }
+    i2c_smbus_write_byte_data(i2c_fd, Registers::MODE1, (mode1_current & ~MODE1_SLEEP) | MODE1_RESTART);
+}
+
+void PCA9685::helloworld() {
+
+}
+
+void PCA9685::setFrequency(uint16_t frequency) {
+    static const uint16_t NUM_COUNTS = 4096; // scaling factor since clock sucks balls lol
+    // if given frequency is not valid, do nothing
+    if (frequency < 24 || frequency > 1526) {
+        return;
+    }
+    output_frequency = frequency;
+    uint8_t prescale = static_cast<uint8_t>(std::round(25e6/(NUM_COUNTS * frequency))) - 1;
```

Comment:
MAGIC NUMBER IN THE CODE!!
Should change to the clock_frequency value already defined in the constructor.  Could also add named vars for the top+bottom valid frequencies